### ISSUE TITLE
post: trying golang

### ISSUE THREAD

**Issue by metaist** [metaist/metaist.com#16]

**Comment by metaist** [comment]
Completed https://go.dev/tour/

- most of the syntax feels pretty familiar
- capitalize first letter to make a member public (no keyword); interesting choice, will take a little typing to get used to
- can initialize variables in `if` and `switch` similar to `for`; variables exist only within scope
- `defer` is nice for all kinds of cleanup (close file, mutex unlock); can modify named return values
- lack of classes first seems weird, but then brilliant (no resolution order wildness)
- have pointers (efficient), but no pointer arithmetic (excellent)
- implicit interfaces-- yes, like they should be
- generics feel a little bolted on (because they are), but make sense
- goroutines are interesting and seem high-performance
- channels seem better than queues

**Comment by metaist** [comment]
https://go.dev/doc/code
- `go install` makes a binary and puts it in `$HOME/go/bin` (hello world is 1.9 MB!)
- `go fmt` uses tabs! (meh, it's the language standard and I can edit my `.editorconfig`)
- `go mod tidy` when you add new dependencies
- `go test` runs the lightweight test framework (How do I get coverage?)

**Comment by metaist** [comment]
https://go.dev/doc/effective_go
- Written in 2009, not updated for all the new stuff, but it's largely conventions, so it's fine.
- `.X()` for the getter of `.x` and `.SetX()` to be its setter.
- interfaces end in `-er` (Reader, Writer, etc.)
- uses `MixedCaps` or `mixedCaps`
- automatic semicolon insertion; no braces on newlines by themselves (yay!)
- `switch` as idiomatic long `if-if else-else`; can have comma-separated lists for cases
- can `break` to a label (like in js)
- "Arrays are values. Assigning one array to another copies all the elements."
- Usually don't pass arrays around, use slices.
- Not sure why I didn't get that runes have single quotes; like `char` in C
- What is `iota`?
  - https://go.dev/wiki/Iota
  - It's a thing that gets reset to zero every time there's `const` and is incremented for every non-blank, non-comment line by one
  - basically an easy way to enumerate enums by using `const`
- there are forms of "type casting" that don't create new values (e.g., `type Sequence []int` can be "cast" to `[]int(s)` because we're just pretending it's a different type)
- you can rename import (I was wondering how that worked)
- you can embed interfaces within other interfaces
- you can make the union of structs by having a pointer to the name of the interface, but not naming the fields
- embedding is somewhat cleaner than subclassing; method calls to the inner get the inner type, not the outer type
- concurrency is not parallelism: https://go.dev/blog/waza-talk
  - ok, but concurrency enabled parallelism
- has a built-in templating language

**Comment by metaist** [comment]
https://go.dev/doc/comment
- kinda like markdown, but a few gotchas

**Comment by metaist** [comment]
https://gobyexample.com/worker-pools
- this makes a lot of sense to me
- the channels really are better than queues

I was surprised this syntax didn't work:

```go
for r := range out {
	fmt.Println("Got result:", r)
}
```

Meaning, it ran until the `out` channel was out of things to do and then crashed.

**Comment by metaist** [comment]
**Are there optional arguments?**
Nope and no overloading either.
https://go.dev/doc/faq#overloading

**Comment by metaist** [comment]
**Is there an equivalent of Python's `range` (formerly `xrange`)?**
Not really. Just use `for`.

**Comment by metaist** [comment]
**Is there an equivalent of `yield`?**
No. Just use channels.
https://go.dev/talks/2013/go4python.slide#20

**Comment by metaist** [comment]
Even cleaner: https://go.dev/talks/2013/go4python.slide#21

**Comment by metaist** [comment]
https://gobyexample.com/

**Comment by metaist** [comment]
Learned a lot; still no real reason for me to use `go` right now.
